Add support for esp32-c3

**Describe the bug**
Latest Binary doesn't run on esp32-c3s. Flashing works but when booting there is always the following repeated error message:
ESP-ROM:esp32c3-api1-20210207
Build:Feb 7 2021
rst:0x10 (TGOWDT_SYS_RST),boot:0xc (SPI_FAST_FLASH_BOOT)
Saved PC:0x40047ed2
Invalid chip_id, Expected 5 read 0. Bootloader for wrong chip?
ets_main.c 333
..

**Expected behavior**
ESPresence boots also on this chip.

**Hardware Details**
ESP32-c3s Development-Board

**Compile on Platform fails**
As this chip seems to use another architecture than other esp32 I tried to compile it myself using platformio in vscode. Sadly it fails and I'm not as skilled in C(++) to be able to fix the errors:

> Executing task in folder ESPresense: platformio run <

Processing esp32c3 (platform: espressif32; framework: arduino; board: esp32dev)
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https://docs.platformio.org/page/boards/espressif32/esp32dev.html
PLATFORM: Espressif 32 (3.4.0) > Espressif ESP32 Dev Module
HARDWARE: ESP32C3 160MHz, 320KB RAM, 4MB Flash
DEBUG: Current (esp-prog) External (esp-prog, iot-bus-jtag, jlink, minimodule, olimex-arm-usb-ocd, olimex-arm-usb-ocd-h, olimex-arm-usb-tiny-h, olimex-jtag-tiny, tumpa)
PACKAGES:
 - framework-arduinoespressif32 0.0.0+sha.399f4ec
 - tool-esptoolpy 1.30100.210531 (3.1.0)
 - toolchain-riscv-esp 1.80400.0 (8.4.0)
LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ chain, Compatibility ~ soft
Found 42 compatible libraries
Scanning dependencies...
Dependency Graph
|-- <SoftFilters> 0.1.0
|-- <AsyncMqttClient> 0.9.0
|   |-- <AsyncTCP> 1.1.1
|-- <ArduinoJson> 6.18.5
|-- <ESP-WiFiSettings> 3.8.0+sha.9a45666
|   |-- <WebServer> 2.0.0
|   |   |-- <WiFi> 2.0.0
|   |   |-- <FS> 2.0.0
|   |-- <DNSServer> 1.1.0
|   |   |-- <WiFi> 2.0.0
|   |-- <LittleFS> 2.0.0
|   |   |-- <FS> 2.0.0
|   |-- <SPIFFS> 2.0.0
|   |   |-- <FS> 2.0.0
|   |-- <WiFi> 2.0.0
|-- <NimBLE-Arduino> 1.3.1+sha.245eefc
|-- <AsyncTCP> 1.1.1
|-- <DNSServer> 1.1.0
|   |-- <WiFi> 2.0.0
|-- <Adafruit Unified Sensor> 1.1.4
|-- <DHT sensor library for ESPx> 1.18.0
|-- <hp_BH1750> 1.0.0
|   |-- <Wire> 2.0.0
|-- <BleFingerprint>
|   |-- <NimBLE-Arduino> 1.3.1+sha.245eefc
|   |-- <ArduinoJson> 6.18.5
|   |-- <SoftFilters> 0.1.0
|-- <ArduinoOTA> 2.0.0
|   |-- <Update> 2.0.0
|   |-- <WiFi> 2.0.0
|   |-- <ESPmDNS> 2.0.0
|   |   |-- <WiFi> 2.0.0
|-- <HTTPClient> 2.0.0
|   |-- <WiFi> 2.0.0
|   |-- <WiFiClientSecure> 2.0.0
|   |   |-- <WiFi> 2.0.0
|-- <HTTPUpdate> 2.0.0
|   |-- <HTTPClient> 2.0.0
|   |   |-- <WiFi> 2.0.0
|   |   |-- <WiFiClientSecure> 2.0.0
|   |   |   |-- <WiFi> 2.0.0
|   |-- <Update> 2.0.0
|   |-- <WiFi> 2.0.0
|-- <SPIFFS> 2.0.0
|   |-- <FS> 2.0.0
|-- <Ticker> 2.0.0
|-- <WebServer> 2.0.0
|   |-- <WiFi> 2.0.0
|   |-- <FS> 2.0.0
|-- <WiFi> 2.0.0
|-- <WiFiClientSecure> 2.0.0
|   |-- <WiFi> 2.0.0
|-- <Wire> 2.0.0
Building in release mode
Compiling .pio/build/esp32c3/src/main.cpp.o
Generating partitions .pio/build/esp32c3/partitions.bin
Compiling .pio/build/esp32c3/lib338/AsyncTCP/AsyncTCP.cpp.o
In file included from /home/dei/.platformio/packages/toolchain-riscv-esp/riscv32-esp-elf/sys-include/string.h:24,
                 from /home/dei/.platformio/packages/framework-arduinoespressif32@src-cba3def1496a47e6af73c0b73bd2e13c/cores/esp32/Arduino.h:29,
                 from lib/BleFingerprint/GUI.h:5,
                 from lib/BleFingerprint/BleFingerprint.h:4,
                 from src/main.h:1,
                 from src/main.cpp:1:
lib/BleFingerprint/strings.h:6:6: error: 'string' in namespace 'std' does not name a type
 std::string slugify(std::string text);
      ^~~~~~
lib/BleFingerprint/strings.h:6:1: note: 'std::string' is defined in header '<string>'; did you forget to '#include <string>'?
+#include <string>

 std::string slugify(std::string text);
 ^~~
lib/BleFingerprint/strings.h:7:1: error: 'String' does not name a type; did you mean 'Sprintf'?
 String slugify(String text);
 ^~~~~~
 Sprintf
lib/BleFingerprint/strings.h:8:6: error: 'string' in namespace 'std' does not name a type
 std::string kebabify(std::string text);
      ^~~~~~
lib/BleFingerprint/strings.h:8:1: note: 'std::string' is defined in header '<string>'; did you forget to '#include <string>'?
 std::string kebabify(std::string text);
 ^~~
lib/BleFingerprint/strings.h:9:1: error: 'String' does not name a type; did you mean 'Sprintf'?
 String kebabify(String text);
 ^~~~~~
 Sprintf
lib/BleFingerprint/strings.h:10:6: error: 'string' in namespace 'std' does not name a type
 std::string hexStr(const char *data, int len);
      ^~~~~~
lib/BleFingerprint/strings.h:10:1: note: 'std::string' is defined in header '<string>'; did you forget to '#include <string>'?
 std::string hexStr(const char *data, int len);
 ^~~
lib/BleFingerprint/strings.h:11:6: error: 'string' in namespace 'std' does not name a type
 std::string hexStr(std::string s);
      ^~~~~~
lib/BleFingerprint/strings.h:11:1: note: 'std::string' is defined in header '<string>'; did you forget to '#include <string>'?
 std::string hexStr(std::string s);
 ^~~
Compiling .pio/build/esp32c3/libc74/AsyncMqttClient/AsyncMqttClient.cpp.o
Compiling .pio/build/esp32c3/libc74/AsyncMqttClient/AsyncMqttClient/Packets/ConnAckPacket.cpp.o
In file included from src/main.cpp:1:
src/main.h: In function 'String resetReason(RESET_REASON)':
src/main.h:107:10: error: 'SW_RESET' was not declared in this scope
     case SW_RESET:
          ^~~~~~~~
src/main.h:107:10: note: suggested alternative: 'SO_REUSE'
     case SW_RESET:
          ^~~~~~~~
          SO_REUSE
src/main.h:109:10: error: 'OWDT_RESET' was not declared in this scope
     case OWDT_RESET:
          ^~~~~~~~~~
src/main.h:109:10: note: suggested alternative: 'ENETRESET'
     case OWDT_RESET:
          ^~~~~~~~~~
          ENETRESET
src/main.h:113:10: error: 'SDIO_RESET' was not declared in this scope
     case SDIO_RESET:
          ^~~~~~~~~~
src/main.h:113:10: note: suggested alternative: 'SO_REUSE'
     case SDIO_RESET:
          ^~~~~~~~~~
          SO_REUSE
src/main.h:123:10: error: 'TGWDT_CPU_RESET' was not declared in this scope
     case TGWDT_CPU_RESET:
          ^~~~~~~~~~~~~~~
src/main.h:123:10: note: suggested alternative: 'TG1WDT_CPU_RESET'
     case TGWDT_CPU_RESET:
          ^~~~~~~~~~~~~~~
          TG1WDT_CPU_RESET
src/main.h:125:10: error: 'SW_CPU_RESET' was not declared in this scope
     case SW_CPU_RESET:
          ^~~~~~~~~~~~
src/main.h:125:10: note: suggested alternative: 'RTC_SW_CPU_RESET'
     case SW_CPU_RESET:
          ^~~~~~~~~~~~
          RTC_SW_CPU_RESET
src/main.h:129:10: error: 'EXT_CPU_RESET' was not declared in this scope
     case EXT_CPU_RESET:
          ^~~~~~~~~~~~~
src/main.h:129:10: note: suggested alternative: 'TG1WDT_CPU_RESET'
     case EXT_CPU_RESET:
          ^~~~~~~~~~~~~
          TG1WDT_CPU_RESET
In file included from src/main.h:5,
                 from src/main.cpp:1:
src/main.h: In function 'void firmwareUpdate()':
src/main.h:219:110: error: 'FIRMWARE' was not declared in this scope
     String firmwareUrl = Sprintf("https://github.com/ESPresense/ESPresense/releases/latest/download/%s.bin", FIRMWARE);
                                                                                                              ^~~~~~~~
src/defaults.h:1:53: note: in definition of macro 'Sprintf'
 #define Sprintf(f, ...) ({ char* s; asprintf(&s, f, __VA_ARGS__); String r = s; free(s); r; })
                                                     ^~~~~~~~~~~
src/main.h:219:110: note: suggested alternative: 'FMARK'
     String firmwareUrl = Sprintf("https://github.com/ESPresense/ESPresense/releases/latest/download/%s.bin", FIRMWARE);
                                                                                                              ^~~~~~~~
src/defaults.h:1:53: note: in definition of macro 'Sprintf'
 #define Sprintf(f, ...) ({ char* s; asprintf(&s, f, __VA_ARGS__); String r = s; free(s); r; })
                                                     ^~~~~~~~~~~
In file included from src/main.cpp:1:
src/main.h: In function 'void commonDiscovery(ArduinoJson6185_91::JsonDocument*)':
src/main.h:316:51: error: expected ';' before 'FIRMWARE'
     (*doc)["dev"]["manufacturer"] = "ESPresense (" FIRMWARE ")";
                                                   ^~~~~~~~~
                                                   ;
src/main.h: In function 'bool sendSwitchDiscovery(String, String)':
src/main.h:459:17: error: 'slugify' was not declared in this scope
     auto slug = slugify(name);
                 ^~~~~~~
src/main.h:459:17: note: suggested alternative: 'slug'
     auto slug = slugify(name);
                 ^~~~~~~
                 slug
src/main.h: In function 'bool sendNumberDiscovery(String, String)':
src/main.h:487:17: error: 'slugify' was not declared in this scope
     auto slug = slugify(name);
                 ^~~~~~~
src/main.h:487:17: note: suggested alternative: 'slug'
     auto slug = slugify(name);
                 ^~~~~~~
                 slug
src/main.cpp: In function 'bool sendTelemetry(int, int, int, int)':
src/main.cpp:39:27: error: 'FIRMWARE' was not declared in this scope
     tele["firm"] = String(FIRMWARE);
                           ^~~~~~~~
src/main.cpp:39:27: note: suggested alternative: 'FMARK'
     tele["firm"] = String(FIRMWARE);
                           ^~~~~~~~
                           FMARK
src/main.cpp: In function 'void connectToWifi()':
src/main.cpp:148:45: error: 'kebabify' was not declared in this scope
     WiFiSettings.hostname = "espresense-" + kebabify(room);
                                             ^~~~~~~~
src/main.cpp:185:10: error: 'slugify' was not declared in this scope
     id = slugify(room);
          ^~~~~~~
Compiling .pio/build/esp32c3/libc74/AsyncMqttClient/AsyncMqttClient/Packets/Out/Connect.cpp.o
*** [.pio/build/esp32c3/src/main.cpp.o] Error 1
================================================================================= [FAILED] Took 9.38 seconds =================================================================================

Environment    Status    Duration
-------------  --------  ------------
esp32c3        FAILED    00:00:09.375
=========================================================================== 1 failed, 0 succeeded in 00:00:09.375 ===========================================================================
The terminal process terminated with exit code: 1

Terminal will be reused by tasks, press any key to close it.

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -37,7 +37,7 @@ jobs:
       run: echo "PLATFORMIO_BUILD_FLAGS=-DVERSION='\"${GITHUB_REF#refs/*/}\"'" >> $GITHUB_ENV
       if: startsWith(github.ref, 'refs/tags/v')
     - name: Run PlatformIO
-      run: pio run -e esp32 -e esp32-verbose -e adafruit-huzzah32 -e m5stickc -e m5stickc-plus -e m5atom-matrix -e esp32-sensors -e esp32-verbose-sensors -e adafruit-huzzah32-sensors -e m5stickc-sensors -e m5stickc-plus-sensors -e m5atom-matrix-sensors -e macchina-a0
+      run: pio run -e esp32 -e esp32-verbose -e adafruit-huzzah32 -e m5stickc -e m5stickc-plus -e m5atom-matrix -e esp32-sensors -e esp32-verbose-sensors -e adafruit-huzzah32-sensors -e m5stickc-sensors -e m5stickc-plus-sensors -e m5atom-matrix-sensors -e macchina-a0 -e esp32c3
     - name: Rename firmware
       run: |
         cp .pio/build/esp32/firmware.bin esp32.bin
@@ -54,6 +54,7 @@ jobs:
         cp .pio/build/m5stickc-plus-sensors/firmware.bin m5stickc-plus-sensors.bin
         cp .pio/build/m5atom-matrix-sensors/firmware.bin m5atom-matrix-sensors.bin
 
+        cp .pio/build/esp32c3/firmware.bin esp32c3.bin
         cp .pio/build/macchina-a0/firmware.bin macchina-a0.bin
     - name: Release
       uses: softprops/action-gh-release@v1

--- a/.gitignore
+++ b/.gitignore
@@ -13,3 +13,4 @@ cmake-build-*
 build/
 CMakeLists.txt
 CMakeListsPrivate.txt
+platformio_override.ini

--- a/lib/BleFingerprint/BleFingerprint.cpp
+++ b/lib/BleFingerprint/BleFingerprint.cpp
@@ -298,7 +298,7 @@ void BleFingerprint::fingerprintManufactureData(NimBLEAdvertisedDevice *advertis
             }
             else if (strManufacturerData.length() >= 4 && strManufacturerData[2] == 0x10)
             {
-                String pid = Sprintf("apple:%02x%02x:%zu", strManufacturerData[2], strManufacturerData[3], strManufacturerData.length());
+                String pid = Sprintf("apple:%02x%02x:%u", strManufacturerData[2], strManufacturerData[3], strManufacturerData.length());
                 if (haveTxPower) pid += -txPower;
                 setId(pid, ID_TYPE_APPLE_NEARBY);
                 disc = hexStr(strManufacturerData.substr(4)).c_str();
@@ -306,7 +306,7 @@ void BleFingerprint::fingerprintManufactureData(NimBLEAdvertisedDevice *advertis
             }
             else if (strManufacturerData.length() >= 4)
             {
-                String pid = Sprintf("apple:%02x%02x:%zu", strManufacturerData[2], strManufacturerData[3], strManufacturerData.length());
+                String pid = Sprintf("apple:%02x%02x:%u", strManufacturerData[2], strManufacturerData[3], strManufacturerData.length());
                 if (haveTxPower) pid += -txPower;
                 setId(pid, ID_TYPE_MISC_APPLE);
                 mdRssi = BleFingerprintCollection::refRssi + APPLE_TX;
@@ -352,7 +352,7 @@ void BleFingerprint::fingerprintManufactureData(NimBLEAdvertisedDevice *advertis
         else if (manuf != "0000")
         {
             mdRssi = haveTxPower ? BleFingerprintCollection::refRssi + txPower : NO_RSSI;
-            String fingerprint = Sprintf("md:%s:%zu", manuf.c_str(), strManufacturerData.length());
+            String fingerprint = Sprintf("md:%s:%u", manuf.c_str(), strManufacturerData.length());
             if (haveTxPower) fingerprint = fingerprint + String(-txPower);
             setId(fingerprint, ID_TYPE_MD);
         }

--- a/lib/BleFingerprint/BleFingerprint.cpp
+++ b/lib/BleFingerprint/BleFingerprint.cpp
@@ -1,7 +1,7 @@
 #include "BleFingerprint.h"
 #include "BleFingerprintCollection.h"
 #include "rssi.h"
-#include "strings.h"
+#include "string_utils.h"
 #include "util.h"
 
 class ClientCallbacks : public BLEClientCallbacks

--- a/lib/BleFingerprint/BleFingerprint.h
+++ b/lib/BleFingerprint/BleFingerprint.h
@@ -3,7 +3,7 @@
 
 #include "GUI.h"
 #include "rssi.h"
-#include "strings.h"
+#include "string_utils.h"
 #include <ArduinoJson.h>
 #include <NimBLEAdvertisedDevice.h>
 #include <NimBLEBeacon.h>

--- a/lib/BleFingerprint/string_utils.cpp
+++ b/lib/BleFingerprint/string_utils.cpp
@@ -1,5 +1,5 @@
 #include <regex>
-#include <strings.h>
+#include <string_utils.h>
 
 std::string ltrim(const std::string &s, char toTrim)
 {

--- a/lib/BleFingerprint/string_utils.h
+++ b/lib/BleFingerprint/string_utils.h
@@ -1,9 +1,15 @@
-#ifndef STRINGS_h
-#define STRINGS_h
+#pragma  once
 
 #include <Arduino.h>
+#include <WString.h>
+#include <string>
+#include <cstring>
 
+#ifdef CONFIG_IDF_TARGET_ESP32C3
+#define ESPMAC (Sprintf("%06x", (uint32_t)ESP.getEfuseMac()))
+#else
 #define ESPMAC (Sprintf("%06" PRIx64, ESP.getEfuseMac() >> 24))
+#endif
 #define Sprintf(f, ...) ({ char* s; asprintf(&s, f, __VA_ARGS__); String r = s; free(s); r; })
 #define Stdprintf(f, ...) ({ char* s; asprintf(&s, f, __VA_ARGS__); std::string r = s; free(s); r; })
 #define SMacf(f) (                                                                                                                                       \
@@ -23,4 +29,3 @@ std::string hexStrRev(const uint8_t *data, int len);
 std::string hexStrRev(const char *data, int len);
 std::string hexStrRev(const std::string &s);
 bool prefixExists(const String& prefixes, const String& s);
-#endif

--- a/platformio.ini
+++ b/platformio.ini
@@ -43,8 +43,9 @@ platform_packages = framework-arduinoespressif32 @ https://github.com/Aircoookie
 framework = ${common.framework}
 build_unflags = ${common.build_unflags}
 build_flags = -g
-  -DARDUINO_ARCH_ESP32
-  -DCONFIG_BT_NIMBLE_PINNED_TO_CORE=1
+  -D ARDUINO_ARCH_ESP32
+  -D CONFIG_BT_NIMBLE_PINNED_TO_CORE=1
+  -D REPORT_PINNED_TO_CORE=1
   ${common.build_flags}
 lib_deps = ${common.lib_deps}
 
@@ -54,10 +55,12 @@ platform_packages =
 framework = ${common.framework}
 build_unflags = ${common.build_unflags}
 build_flags = -g
-  -DARDUINO_ARCH_ESP32
-  -DARDUINO_ARCH_ESP32C3
-  -DCONFIG_IDF_TARGET_ESP32C3
-  -DCO
+  -D ARDUINO_ARCH_ESP32
+  -D ARDUINO_ARCH_ESP32C3
+  -D CONFIG_IDF_TARGET_ESP32C3
+  -D CO
+  -D CONFIG_BT_NIMBLE_PINNED_TO_CORE=0
+  -D REPORT_PINNED_TO_CORE=0
   ${common.build_flags}
 lib_deps = ${common.lib_deps}
 

--- a/platformio.ini
+++ b/platformio.ini
@@ -10,6 +10,8 @@
 
 [platformio]
 default_envs = esp32
+build_cache_dir = ~/.buildcache
+extra_configs = platformio_override.ini
 
 [common]
 debug_build_flags = -O0 -ggdb3 -g3 -DDEBUG_TLS_MEM
@@ -17,7 +19,6 @@ build_flags =
 	-D MQTT_MIN_FREE_MEMORY=8128
 	-D SECURE_CLIENT=SECURE_CLIENT_BEARSSL
 	-D BEARSSL_SSL_BASIC
-	-D CONFIG_BT_NIMBLE_PINNED_TO_CORE=1
 	-D CONFIG_BT_NIMBLE_MAX_BONDS=0
 	-D CONFIG_BT_NIMBLE_MAX_CCCDS=0
 	-D CONFIG_BT_NIMBLE_TASK_STACK_SIZE=8128
@@ -25,7 +26,7 @@ build_flags =
 	-D CONFIG_BT_NIMBLE_ROLE_BROADCASTER_DISABLED
 	-D CONFIG_ASYNC_TCP_USE_WDT=0
 	-D CONFIG_USE_ETHERNET
-platform = espressif32@3.2
+build_unflags =
 framework = arduino
 lib_deps =
 	haimoz/SoftFilters@^0.1.0
@@ -35,6 +36,30 @@ lib_deps =
 	https://github.com/ESPresense/NimBLE-Arduino.git
 	https://github.com/pbolduc/AsyncTCP.git @ 1.2.0
 	bbx10/DNSServer@^1.1.0
+
+[esp32]
+platform = espressif32@3.5.0
+platform_packages = framework-arduinoespressif32 @ https://github.com/Aircoookie/arduino-esp32.git#1.0.6.4
+framework = ${common.framework}
+build_unflags = ${common.build_unflags}
+build_flags = -g
+  -DARDUINO_ARCH_ESP32
+  -DCONFIG_BT_NIMBLE_PINNED_TO_CORE=1
+  ${common.build_flags}
+lib_deps = ${common.lib_deps}
+
+[esp32c3]
+platform = https://github.com/tasmota/platform-espressif32/releases/download/v2.0.2.2/platform-tasmota-espressif32-2.0.2.zip
+platform_packages =
+framework = ${common.framework}
+build_unflags = ${common.build_unflags}
+build_flags = -g
+  -DARDUINO_ARCH_ESP32
+  -DARDUINO_ARCH_ESP32C3
+  -DCONFIG_IDF_TARGET_ESP32C3
+  -DCO
+  ${common.build_flags}
+lib_deps = ${common.lib_deps}
 
 [common_sensors]
 lib_deps =
@@ -46,225 +71,267 @@ lib_deps =
 
 [env:debug]
 build_type=debug
-platform = ${common.platform}
-framework = ${common.framework}
+platform = ${esp32.platform}
+platform_packages = ${esp32.platform_packages}
+framework = ${esp32.framework}
 board = esp-wrover-kit
 debug_tool = esp-prog
 debug_init_break=tbreak abort
 upload_protocol = esp-prog
-lib_deps = ${common.lib_deps}
+lib_deps = ${esp32.lib_deps}
 board_build.partitions = partitions_singleapp.csv
 monitor_port = /dev/cu.usbserial-*1
 monitor_speed = 115200
 monitor_filters = esp32_exception_decoder, time
 debug_build_flags = ${common.debug_build_flags}
-build_flags = ${common.build_flags}
+build_unflags = ${esp32.build_unflags}
+build_flags = ${esp32.build_flags}
 
 [env:esp32]
-platform = ${common.platform}
-framework = ${common.framework}
+platform = ${esp32.platform}
+platform_packages = ${esp32.platform_packages}
+framework = ${esp32.framework}
 board = esp32dev
-lib_deps = ${common.lib_deps}
+lib_deps = ${esp32.lib_deps}
 board_build.partitions = partitions_singleapp.csv
 monitor_speed = 115200
 upload_speed = 1500000
 monitor_filters = esp32_exception_decoder
+build_unflags = ${esp32.build_unflags}
 build_flags =
 	-D FIRMWARE='"esp32"'
-	${common.build_flags}
+	${esp32.build_flags}
+
+[env:esp32c3]
+platform = ${esp32c3.platform}
+platform_packages = ${esp32c3.platform_packages}
+framework = ${esp32c3.framework}
+board = esp32-c3-devkitm-1
+board_build.partitions = partitions_singleapp.csv
+monitor_speed = 115200
+upload_speed = 1500000
+lib_deps = ${esp32c3.lib_deps}
+build_unflags = ${esp32c3.build_unflags}
+build_flags =
+	-D FIRMWARE='"esp32c3"'
+	${esp32c3.build_flags}
 
 [env:adafruit-huzzah32]
-platform = ${common.platform}
-framework = ${common.framework}
+platform = ${esp32.platform}
+platform_packages = ${esp32.platform_packages}
+framework = ${esp32.framework}
 board = esp32dev
-lib_deps = ${common.lib_deps}
+lib_deps = ${esp32.lib_deps}
 board_build.partitions = partitions_singleapp.csv
 monitor_speed = 115200
 monitor_filters = esp32_exception_decoder
+build_unflags = ${esp32.build_unflags}
 build_flags =
 	-D FIRMWARE='"adafruit-huzzah32"'
 	-D HUZZAH32
-	${common.build_flags}
+	${esp32.build_flags}
 
 [env:esp32-verbose]
-platform = ${common.platform}
-framework = ${common.framework}
+platform = ${esp32.platform}
+platform_packages = ${esp32.platform_packages}
+framework = ${esp32.framework}
 board = esp32dev
-lib_deps = ${common.lib_deps}
+lib_deps = ${esp32.lib_deps}
 board_build.partitions = partitions_singleapp.csv
 monitor_speed = 115200
 upload_speed = 1500000
 monitor_filters = esp32_exception_decoder
+build_unflags = ${esp32.build_unflags}
 build_flags =
 	-D CORE_DEBUG_LEVEL=2
 	-D FIRMWARE='"esp32-verbose"'
 	-D VERBOSE
-	${common.build_flags}
+	${esp32.build_flags}
 
 [env:m5stickc]
-platform = ${common.platform}
-framework = ${common.framework}
+platform = ${esp32.platform}
+platform_packages = ${esp32.platform_packages}
+framework = ${esp32.framework}
 board = m5stick-c
 lib_deps =
 	m5stack/M5StickC@^0.2.0
 	https://github.com/ESPresense/M5StickC-TB_Display.git
-	${common.lib_deps}
+	${esp32.lib_deps}
 board_build.partitions = partitions_singleapp.csv
 monitor_speed = 115200
 monitor_filters = esp32_exception_decoder
+build_unflags = ${esp32.build_unflags}
 build_flags =
 	-D M5STICK
 	-D FIRMWARE='"m5stickc"'
-	${common.build_flags}
+	${esp32.build_flags}
 
 [env:m5stickc-plus]
-platform = ${common.platform}
-framework = ${common.framework}
+platform = ${esp32.platform}
+platform_packages = ${esp32.platform_packages}
+framework = ${esp32.framework}
 board = m5stick-c
 lib_deps =
 	m5stack/M5StickCPlus@^0.0.2
 	https://github.com/ESPresense/M5StickC-TB_Display.git
-	${common.lib_deps}
+	${esp32.lib_deps}
 board_build.partitions = partitions_singleapp.csv
 monitor_speed = 115200
 monitor_filters = esp32_exception_decoder
+build_unflags = ${esp32.build_unflags}
 build_flags =
 	-D M5STICK
 	-D PLUS
 	-D FIRMWARE='"m5stickc-plus"'
-	${common.build_flags}
+	${esp32.build_flags}
 
 [env:m5atom-matrix]
-platform = ${common.platform}
-framework = ${common.framework}
+platform = ${esp32.platform}
+platform_packages = ${esp32.platform_packages}
+framework = ${esp32.framework}
 board = m5stack-atom
 lib_deps =
 	fastled/FastLED@^3.4.0
 	m5stack/m5atom@^0.0.5
-	${common.lib_deps}
+	${esp32.lib_deps}
 board_build.partitions = partitions_singleapp.csv
 monitor_speed = 115200
 monitor_filters = esp32_exception_decoder
+build_unflags = ${esp32.build_unflags}
 build_flags =
 	-D M5ATOM
 	-D MATRIX
 	-D FIRMWARE='"m5atom-matrix"'
-	${common.build_flags}
+	${esp32.build_flags}
 
 [env:macchina-a0]
-platform = ${common.platform}
-framework = ${common.framework}
+platform = ${esp32.platform}
+platform_packages = ${esp32.platform_packages}
+framework = ${esp32.framework}
 board = esp32dev
-lib_deps = ${common.lib_deps}
+lib_deps = ${esp32.lib_deps}
 board_build.partitions = partitions_singleapp.csv
 monitor_speed = 115200
+build_unflags = ${esp32.build_unflags}
 build_flags =
 	-D MACCHINA_A0
 	-D FIRMWARE='"macchina-a0"'
-	${common.build_flags}
+	${esp32.build_flags}
 
 [env:esp32-sensors]
-platform = ${common.platform}
-framework = ${common.framework}
+platform = ${esp32.platform}
+platform_packages = ${esp32.platform_packages}
+framework = ${esp32.framework}
 board = esp32dev
 lib_deps =
-	${common.lib_deps}
+	${esp32.lib_deps}
 	${common_sensors.lib_deps}
 board_build.partitions = partitions_singleapp.csv
 monitor_speed = 115200
 upload_speed = 115200
 monitor_filters = esp32_exception_decoder
+build_unflags = ${esp32.build_unflags}
 build_flags =
 	-D FIRMWARE='"esp32-sensors"'
 	-D SENSORS
-	${common.build_flags}
+	${esp32.build_flags}
 
 [env:adafruit-huzzah32-sensors]
-platform = ${common.platform}
-framework = ${common.framework}
+platform = ${esp32.platform}
+platform_packages = ${esp32.platform_packages}
+framework = ${esp32.framework}
 board = esp32dev
 lib_deps =
-	${common.lib_deps}
+	${esp32.lib_deps}
 	${common_sensors.lib_deps}
 board_build.partitions = partitions_singleapp.csv
 monitor_speed = 115200
 monitor_filters = esp32_exception_decoder
+build_unflags = ${esp32.build_unflags}
 build_flags =
 	-D FIRMWARE='"adafruit-huzzah32-sensors"'
 	-D HUZZAH32
 	-D SENSORS
-	${common.build_flags}
+	${esp32.build_flags}
 
 [env:esp32-verbose-sensors]
-platform = ${common.platform}
-framework = ${common.framework}
+platform = ${esp32.platform}
+platform_packages = ${esp32.platform_packages}
+framework = ${esp32.framework}
 board = esp32dev
 lib_deps =
-	${common.lib_deps}
+	${esp32.lib_deps}
 	${common_sensors.lib_deps}
 board_build.partitions = partitions_singleapp.csv
 monitor_speed = 115200
 upload_speed = 1500000
 monitor_filters = esp32_exception_decoder
+build_unflags = ${esp32.build_unflags}
 build_flags =
 	-D CORE_DEBUG_LEVEL=2
 	-D FIRMWARE='"esp32-verbose-sensors"'
 	-D VERBOSE
 	-D SENSORS
-	${common.build_flags}
+	${esp32.build_flags}
 
 [env:m5stickc-sensors]
-platform = ${common.platform}
-framework = ${common.framework}
+platform = ${esp32.platform}
+platform_packages = ${esp32.platform_packages}
+framework = ${esp32.framework}
 board = m5stick-c
 lib_deps =
 	m5stack/M5StickC@^0.2.0
 	https://github.com/ESPresense/M5StickC-TB_Display.git
-	${common.lib_deps}
+	${esp32.lib_deps}
 	${common_sensors.lib_deps}
 board_build.partitions = partitions_singleapp.csv
 monitor_speed = 115200
 monitor_filters = esp32_exception_decoder
+build_unflags = ${esp32.build_unflags}
 build_flags =
 	-D M5STICK
 	-D FIRMWARE='"m5stickc-sensors"'
 	-D SENSORS
-	${common.build_flags}
+	${esp32.build_flags}
 
 [env:m5stickc-plus-sensors]
-platform = ${common.platform}
-framework = ${common.framework}
+platform = ${esp32.platform}
+platform_packages = ${esp32.platform_packages}
+framework = ${esp32.framework}
 board = m5stick-c
 lib_deps =
 	m5stack/M5StickCPlus@^0.0.2
 	https://github.com/ESPresense/M5StickC-TB_Display.git
-	${common.lib_deps}
+	${esp32.lib_deps}
 	${common_sensors.lib_deps}
 board_build.partitions = partitions_singleapp.csv
 monitor_speed = 115200
 monitor_filters = esp32_exception_decoder
+build_unflags = ${esp32.build_unflags}
 build_flags =
 	-D M5STICK
 	-D PLUS
 	-D FIRMWARE='"m5stickc-plus-sensors"'
 	-D SENSORS
-	${common.build_flags}
+	${esp32.build_flags}
 
 [env:m5atom-matrix-sensors]
-platform = ${common.platform}
-framework = ${common.framework}
+platform = ${esp32.platform}
+platform_packages = ${esp32.platform_packages}
+framework = ${esp32.framework}
 board = m5stack-atom
 lib_deps =
 	fastled/FastLED@^3.4.0
 	m5stack/m5atom@^0.0.5
-	${common.lib_deps}
+	${esp32.lib_deps}
 	${common_sensors.lib_deps}
 board_build.partitions = partitions_singleapp.csv
 monitor_speed = 115200
 monitor_filters = esp32_exception_decoder
+build_unflags = ${esp32.build_unflags}
 build_flags =
 	-D M5ATOM
 	-D MATRIX
 	-D FIRMWARE='"m5atom-matrix-sensors"'
 	-D SENSORS
-	${common.build_flags}
+	${esp32.build_flags}

--- a/platformio.ini
+++ b/platformio.ini
@@ -25,7 +25,6 @@ build_flags =
 	-D CONFIG_BT_NIMBLE_ROLE_PERIPHERAL_DISABLED
 	-D CONFIG_BT_NIMBLE_ROLE_BROADCASTER_DISABLED
 	-D CONFIG_ASYNC_TCP_USE_WDT=0
-	-D CONFIG_USE_ETHERNET
 build_unflags =
 framework = arduino
 lib_deps =
@@ -46,11 +45,13 @@ build_flags = -g
   -D ARDUINO_ARCH_ESP32
   -D CONFIG_BT_NIMBLE_PINNED_TO_CORE=1
   -D REPORT_PINNED_TO_CORE=1
+  -D CONFIG_USE_ETHERNET
+
   ${common.build_flags}
 lib_deps = ${common.lib_deps}
 
 [esp32c3]
-platform = https://github.com/tasmota/platform-espressif32/releases/download/v2.0.2.2/platform-tasmota-espressif32-2.0.2.zip
+platform = espressif32@4.4.0
 platform_packages =
 framework = ${common.framework}
 build_unflags = ${common.build_unflags}
@@ -71,24 +72,6 @@ lib_deps =
 	starmbi/hp_BH1750 @ ^1.0.0
 	adafruit/Adafruit BME280 Library@^2.2.2
 	adafruit/Adafruit TSL2561@^1.1.0
-
-[env:debug]
-build_type=debug
-platform = ${esp32.platform}
-platform_packages = ${esp32.platform_packages}
-framework = ${esp32.framework}
-board = esp-wrover-kit
-debug_tool = esp-prog
-debug_init_break=tbreak abort
-upload_protocol = esp-prog
-lib_deps = ${esp32.lib_deps}
-board_build.partitions = partitions_singleapp.csv
-monitor_port = /dev/cu.usbserial-*1
-monitor_speed = 115200
-monitor_filters = esp32_exception_decoder, time
-debug_build_flags = ${common.debug_build_flags}
-build_unflags = ${esp32.build_unflags}
-build_flags = ${esp32.build_flags}
 
 [env:esp32]
 platform = ${esp32.platform}

--- a/src/BME280Sensor.cpp
+++ b/src/BME280Sensor.cpp
@@ -7,7 +7,7 @@
 #include <Adafruit_BME280.h>
 
 // for #define ESPMAC
-#include "strings.h"
+#include "string_utils.h"
 
 // TODO: Not a fan of externs, but this helps refactoring for now
 extern bool I2C_Bus_1_Enabled;
@@ -30,7 +30,7 @@ namespace BME280
 
     void Setup()
     {
-        
+
     }
 
     void ConnectToWifi()

--- a/src/HX711.cpp
+++ b/src/HX711.cpp
@@ -6,7 +6,7 @@
 #include <AsyncMqttClient.h>
 
 // for #define ESPMAC
-#include "strings.h"
+#include "string_utils.h"
 
 // TODO: Not a fan of externs, but this helps refactoring for now
 extern char buffer[2048];

--- a/src/MotionSensors.cpp
+++ b/src/MotionSensors.cpp
@@ -3,7 +3,7 @@
 #include <AsyncMqttClient.h>
 
 // for #define ESPMAC
-#include "strings.h"
+#include "string_utils.h"
 
 // TODO: Not a fan of externs, but this helps refactoring for now
 extern char buffer[2048];

--- a/src/TSL2561Sensor.cpp
+++ b/src/TSL2561Sensor.cpp
@@ -7,7 +7,7 @@
 #include <Adafruit_TSL2561_U.h>
 
 // for #define ESPMAC
-#include "strings.h"
+#include "string_utils.h"
 
 // TODO: Not a fan of externs, but this helps refactoring for now
 extern bool I2C_Bus_1_Enabled;
@@ -29,7 +29,6 @@ namespace TSL2561
 
     void Setup()
     {
-        
     }
 
     void ConnectToWifi()

--- a/src/main.cpp
+++ b/src/main.cpp
@@ -128,20 +128,26 @@ void setupNetwork()
 
     WiFiSettings.onFailure = []()
     {
-        GUI::status("WiFi Portal...");
+        GUI::status("WiFi Failed");
     };
     WiFiSettings.onWaitLoop = []()
     {
         GUI::connecting();
         return 150;
     };
+    static bool inPortal = false;
     WiFiSettings.onPortalWaitLoop = []()
     {
+        if (!inPortal)
+        {
+            inPortal = true;
+            GUI::status("WiFi Portal...");
+        }
+
         if (getUptimeSeconds() > 600)
             ESP.restart();
     };
 
-    GUI::connected(true, false);
 #ifdef VERSION
     WiFiSettings.info("ESPresense Version: " + String(VERSION));
 #endif
@@ -228,6 +234,9 @@ void setupNetwork()
     if (ethernetType > 0) success = Network.connect(ethernetType, 20, WiFiSettings.hostname.c_str());
     if (!success && !WiFiSettings.connect(true, 40))
         ESP.restart();
+
+    GUI::connected(true, false);
+
 #ifdef FIRMWARE
     Serial.println("Firmware:     " + String(FIRMWARE));
 #endif

--- a/src/main.cpp
+++ b/src/main.cpp
@@ -723,7 +723,7 @@ void setup()
     HX711::Setup();
 #endif
     xTaskCreatePinnedToCore(scanTask, "scanTask", 7168, nullptr, 2, &scanTaskHandle, CONFIG_BT_NIMBLE_PINNED_TO_CORE);
-    xTaskCreatePinnedToCore(reportTask, "reportTask", 7168, nullptr, 1, &reportTaskHandle, 1);
+    xTaskCreatePinnedToCore(reportTask, "reportTask", 7168, nullptr, 1, &reportTaskHandle, REPORT_PINNED_TO_CORE);
     configureOTA();
 }
 

--- a/src/main.h
+++ b/src/main.h
@@ -3,7 +3,7 @@
 
 #include "GUI.h"
 #include "defaults.h"
-#include "strings.h"
+#include "string_utils.h"
 
 #include <Arduino.h>
 #include <ArduinoJson.h>
@@ -114,6 +114,9 @@ BleFingerprintCollection fingerprints;
 
 String resetReason(RESET_REASON reason)
 {
+#ifdef ARDUINO_ARCH_ESP32C3
+    return "";
+#else
     switch (reason)
     {
     case POWERON_RESET:
@@ -149,6 +152,7 @@ String resetReason(RESET_REASON reason)
     default:
         return "Unknown";
     }
+#endif
 }
 
 unsigned long getUptimeSeconds()

